fiftyone/core/clips.py

```diff
@@ -406,6 +407,131 @@ def _sync_source_keep_fields(self):
+            raise ValueError(
+                "Trajectory views must be defined by a %s stage; found %s"
+                % (fost.ToTrajectories, type(clips_stage))
+            )
+
+        if _stages is None:
+            _stages = []
+
+        trajectory_stages = self._make_trajectory_stages(
+            source_collection, clips_stage, clips_dataset
+        )
+        _stages = trajectory_stages + _stages
+
+        self._num_trajectory_stages = len(trajectory_stages)
+        self._classification_field = None
+        self._source_collection = source_collection
+        self._clips_stage = clips_stage
+        self._clips_dataset = clips_dataset
+        self.__stages = _stages
+        self.__name = _name
+
+    def __copy__(self):
+        return self.__class__(
+            self._source_collection,
+            deepcopy(self._clips_stage),
+            self._clips_dataset,
+            _stages=deepcopy(self.__stages[self._num_trajectory_stages :]),
+            _name=self.__name,
+        )
+
+    @property
+    def _stages(self):
+        return self.__stages
+
+    @property
+    def _all_stages(self):
+        return (
+            self._source_collection.view()._all_stages
+            + [self._clips_stage]
+            + self.__stages[self._num_trajectory_stages :]
+        )
+
+    @staticmethod
+    def _make_trajectory_stages(source_collection, clips_stage, clips_dataset):
+        field, _ = source_collection._handle_frame_field(clips_stage.field)
+
+        trajectory_stages = []
+
+        #
+        # Exclude all frame-level fields that weren't explicitly requested
+        #
+
+        exclude_fields = set(source_collection.get_frame_field_schema().keys())
+        exclude_fields -= set(source_collection._get_default_frame_fields())
+        exclude_fields.discard(field)
+        if clips_stage.config and clips_stage.config.get("other_fields", None):
+            other_fields = clips_stage.config["other_fields"]
+            if other_fields == True:
+                exclude_fields.clear()
+            else:
+                if etau.is_str(other_fields):
+                    other_fields = [other_fields]
+
+                for other_field in other_fields:
+                    (
+                        _field,
+                        is_frame_field,
+                    ) = source_collection._handle_frame_field(other_field)
+                    if is_frame_field:
+                        exclude_fields.discard(_field)
+
+        if exclude_fields:
+            exclude_stage = fost.ExcludeFields(
+                [source_collection._FRAMES_PREFIX + f for f in exclude_fields]
+            )
+            exclude_stage.validate(clips_dataset)
+            trajectory_stages.append(exclude_stage)
+
+        #
+        # Select correct trajectory
+        #
+
+        filter_stage = fost.FilterLabels(
+            clips_stage.field,
+            (
+                (F("label") == F("$" + field + ".label"))
```

Comment:
Don't fully understand what this is. I'm reading "$" means look at the sample itself not the embedded document.
So does this basically group clips into equal label/index combos?

Might just be my relative ignorance of ViewExpressions

Reply:
yeah what this is doing is filtering the labels in `clips_stage.field` so that only the ones whose `label` attribute match the `field.label` attribute at the sample-level are retained.

In trajectory views, every clip has a sample-level field that contains the unique `(label, index)` for the unique object associated with the clip.